<a href="https://colab.research.google.com/github/Vicrrs/SVM-SupportVectorMachines-/blob/main/SVM_(Support_Vector_Machines).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Importando as Bibliotecas

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt


In [ ]:
dados = pd.read_csv("https://raw.githubusercontent.com/tirthajyoti/Machine-Learning-with-Python/master/Datasets/loan_data.csv")

In [ ]:
dados

credit.policy             purpose  int.rate  installment  \
0                 1  debt_consolidation    0.1189       829.10   
1                 1         credit_card    0.1071       228.22   
2                 1  debt_consolidation    0.1357       366.86   
3                 1  debt_consolidation    0.1008       162.34   
4                 1         credit_card    0.1426       102.92   
...             ...                 ...       ...          ...   
9573              0           all_other    0.1461       344.76   
9574              0           all_other    0.1253       257.70   
9575              0  debt_consolidation    0.1071        97.81   
9576              0    home_improvement    0.1600       351.58   
9577              0  debt_consolidation    0.1392       853.43   

      log.annual.inc    dti  fico  days.with.cr.line  revol.bal  revol.util  \
0          11.350407  19.48   737        5639.958333      28854        52.1   
1          11.082143  14.29   707        2760.000000      33623        76.7   
2          10.373491  11.63   682        4710.000000       3511        25.6   
3          11.350407   8.10   712        2699.958333      33667        73.2   
4          11.299732  14.97   667        4066.000000       4740        39.5   
...              ...    ...   ...                ...        ...         ...   
9573       12.180755  10.39   672       10474.000000     215372        82.1   
9574       11.141862   0.21   722        4380.000000        184         1.1   
9575       10.596635  13.09   687        3450.041667      10036        82.9   
9576       10.819778  19.18   692        1800.000000          0         3.2   
9577       11.264464  16.28   732        4740.000000      37879        57.0   

      inq.last.6mths  delinq.2yrs  pub.rec  not.fully.paid  
0                  0            0        0               0  
1                  0            0        0               0  
2                  1            0        0               0  
3                  1            0        0               0  
4                  0            1        0               0  
...              ...          ...      ...             ...  
9573               2            0        0               1  
9574               5            0        0               1  
9575               8            0        0               1  
9576               5            0        0               1  
9577               6            0        0               1  

[9578 rows x 14 columns]

In [ ]:
dados.head()

credit.policy             purpose  int.rate  installment  log.annual.inc  \
0              1  debt_consolidation    0.1189       829.10       11.350407   
1              1         credit_card    0.1071       228.22       11.082143   
2              1  debt_consolidation    0.1357       366.86       10.373491   
3              1  debt_consolidation    0.1008       162.34       11.350407   
4              1         credit_card    0.1426       102.92       11.299732   

     dti  fico  days.with.cr.line  revol.bal  revol.util  inq.last.6mths  \
0  19.48   737        5639.958333      28854        52.1               0   
1  14.29   707        2760.000000      33623        76.7               0   
2  11.63   682        4710.000000       3511        25.6               1   
3   8.10   712        2699.958333      33667        73.2               1   
4  14.97   667        4066.000000       4740        39.5               0   

   delinq.2yrs  pub.rec  not.fully.paid  
0            0        0               0  
1            0        0               0  
2            0        0               0  
3            0        0               0  
4            1        0               0

# Aplicando Label Encoder

Iremos aplicar o Label Encoder porque o SVM irá se basear em valores numéricos para definir o melhor plano, então ele só sabe trabalhar com variáveis numéricas. Iremos transformar a coluna Purpose em numérica para continuar.

In [ ]:
dados.dtypes

credit.policy          int64
purpose               object
int.rate             float64
installment          float64
log.annual.inc       float64
dti                  float64
fico                   int64
days.with.cr.line    float64
revol.bal              int64
revol.util           float64
inq.last.6mths         int64
delinq.2yrs            int64
pub.rec                int64
not.fully.paid         int64
dtype: object

In [ ]:
dados.dtypes.reset_index()

index        0
0       credit.policy    int64
1             purpose   object
2            int.rate  float64
3         installment  float64
4      log.annual.inc  float64
5                 dti  float64
6                fico    int64
7   days.with.cr.line  float64
8           revol.bal    int64
9          revol.util  float64
10     inq.last.6mths    int64
11        delinq.2yrs    int64
12            pub.rec    int64
13     not.fully.paid    int64

In [ ]:
colunas = dados.dtypes.reset_index()

In [ ]:
colunas.columns

Index(['index', 0], dtype='object')

In [ ]:
colunas[colunas[0] == 'object']['index']

1    purpose
Name: index, dtype: object

In [ ]:
categ_col = colunas[colunas[0] == 'object']['index'].to_list()

In [ ]:
categ_col

['purpose']

In [ ]:
# Fazendo o Label Encoder
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()

In [ ]:
#criando  um coluna purpose encoded
for i in categ_col:
  le.fit(dados[i])
  dados[str(i) + '_encoded'] = le.transform(dados[i])

In [ ]:
dados

credit.policy             purpose  int.rate  installment  \
0                 1  debt_consolidation    0.1189       829.10   
1                 1         credit_card    0.1071       228.22   
2                 1  debt_consolidation    0.1357       366.86   
3                 1  debt_consolidation    0.1008       162.34   
4                 1         credit_card    0.1426       102.92   
...             ...                 ...       ...          ...   
9573              0           all_other    0.1461       344.76   
9574              0           all_other    0.1253       257.70   
9575              0  debt_consolidation    0.1071        97.81   
9576              0    home_improvement    0.1600       351.58   
9577              0  debt_consolidation    0.1392       853.43   

      log.annual.inc    dti  fico  days.with.cr.line  revol.bal  revol.util  \
0          11.350407  19.48   737        5639.958333      28854        52.1   
1          11.082143  14.29   707        2760.000000      33623        76.7   
2          10.373491  11.63   682        4710.000000       3511        25.6   
3          11.350407   8.10   712        2699.958333      33667        73.2   
4          11.299732  14.97   667        4066.000000       4740        39.5   
...              ...    ...   ...                ...        ...         ...   
9573       12.180755  10.39   672       10474.000000     215372        82.1   
9574       11.141862   0.21   722        4380.000000        184         1.1   
9575       10.596635  13.09   687        3450.041667      10036        82.9   
9576       10.819778  19.18   692        1800.000000          0         3.2   
9577       11.264464  16.28   732        4740.000000      37879        57.0   

      inq.last.6mths  delinq.2yrs  pub.rec  not.fully.paid  purpose_encoded  
0                  0            0        0               0                2  
1                  0            0        0               0                1  
2                  1            0        0               0                2  
3                  1            0        0               0                2  
4                  0            1        0               0                1  
...              ...          ...      ...             ...              ...  
9573               2            0        0               1                0  
9574               5            0        0               1                0  
9575               8            0        0               1                2  
9576               5            0        0               1                4  
9577               6            0        0               1                2  

[9578 rows x 15 columns]

In [ ]:
# Eliminando a coluna que nao preciso mais (purpose)
dados = dados.drop('purpose', axis=1)

In [ ]:
dados.head()

credit.policy  int.rate  installment  log.annual.inc    dti  fico  \
0              1    0.1189       829.10       11.350407  19.48   737   
1              1    0.1071       228.22       11.082143  14.29   707   
2              1    0.1357       366.86       10.373491  11.63   682   
3              1    0.1008       162.34       11.350407   8.10   712   
4              1    0.1426       102.92       11.299732  14.97   667   

   days.with.cr.line  revol.bal  revol.util  inq.last.6mths  delinq.2yrs  \
0        5639.958333      28854        52.1               0            0   
1        2760.000000      33623        76.7               0            0   
2        4710.000000       3511        25.6               1            0   
3        2699.958333      33667        73.2               1            0   
4        4066.000000       4740        39.5               0            1   

   pub.rec  not.fully.paid  purpose_encoded  
0        0               0                2  
1        0               0                1  
2        0               0                2  
3        0               0                2  
4        0               0                1

# Implementando o modelo

Variável que queremos prever é a Not Fully Paid (ou seja, se a pessoa não pagou totalmente o impréstimo)

## Separando a base em X (variáveis de entrada) e Y (variável que queremos prever)

In [ ]:
#separando o dataframe em x -> features/variaveis de entrada, y -> variaveis de resposta/target
x = dados.drop('not.fully.paid', axis=1)
y = dados['not.fully.paid']

## Quebrando o dataset em treino e teste - Usando 70% para treino e 30% para teste

In [ ]:
from sklearn.model_selection import train_test_split

x_train, x_test , y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=42)

## Carregando o Modelo

In [ ]:
from sklearn.svm import SVC

# Selecionando qual modelo eu quero treinar, manteremos o kernel linear
# mas importante pesquisar sobre outros tipos de kernel para o SVM (sigmoid, radial  e polynomial)

svm = SVC(kernel='linear')
svm.fit(x_train, y_train)

# Fazendo a predição

y_pred = svm.predict(x_test)


## Avaliando resultado

In [ ]:
from sklearn.metrics import accuracy_score

print(accuracy_score(y_test, y_pred))


0.8274182324286709


## Conclusão

Dataset de emprestismos, vamos prever a probabilidade de uma pessoal pagar ou nao um emprestimo. Conseguimos um resultado de 82%, para a possibilidade de ela pagar o empréstimo.